### PR TITLE
Add registered charity number validation attribute

### PR DESCRIPTION
**Pull request by samuelhaa**
The validator is case insensitive

Correct formats:

- 6 digit number or
- 7 digit number or
- 6 digit number prefixed with SC for example, SC123456

[AB#155184](https://dev.azure.com/thepensionsregulator/0158f35d-cea5-4d9d-adc8-9fc3d3d98793/_workitems/edit/155184)

### CODE REVIEW

File: docs/components/text-input.md [hunk 1]
```diff
@@ -1,14 +1,14 @@
-# Date input
+## Date input
```
Comment (sussexrick) [inline docs/components/text-input.md]: Sorry, I wasn't clear. The heading levels were correct so you should put them back. I was referring to the h1 saying 'Date input' even though it's about the 'Text input' component

File: docs/components/text-input.md [hunk 2]
```diff
@@ -17,3 +17,22 @@ Error messages for text input components are configurable in Umbraco on the sett
...
 ![Settings for a postcode using a text input component](/docs/images/uk-postcode-settings.png)
+
+
+## Registered charity numbers
+
+### Validating registered charity numbers
+
+Use the `[RegisteredCharityNumber]` attribute to validate registered charity numbers using the Text input component. This uses a regular expression rather than a check against a  database, so some false positives may be allowed.
+
+For an example of this in use, see the 'Text input' page in either the ASP.NET and Umbraco example apps.
```
Comment (sussexrick) [inline docs/components/text-input.md:47]: These examples need to be added to the example apps.

File: GovUk.Frontend.AspNetCore.Extensions/Validation/RegisteredCharityNumberAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    public class RegisteredCharityNumberAttribute : RegularExpressionAttribute
+    {
+        private const string SIX_NUMBERS_ONLY = @"(^[0-9]{6}$)";
+        private const string SEVEN_NUMBERS_ONLY = @"(^[0-9]{7}$)";
+        private const string SC_PREFIX_AND_SIX_NUMBERS = @"(^[Ss][Cc][0-9]{6}$)";
+
+        public RegisteredCharityNumberAttribute()
+            : base($@"{SIX_NUMBERS_ONLY}|{SEVEN_NUMBERS_ONLY}|{SC_PREFIX_AND_SIX_NUMBERS}")
+        {
+            ErrorMessage = "Enter a valid registered charity number";
```
Comment (sussexrick) [inline GovUk.Frontend.AspNetCore.Extensions/Validation/RegisteredCharityNumberAttribute.cs]: 6. 'Enter a valid registered charity number' should be 'Enter a registered charity number in the correct format, like 123456, SC123456 or 1234567'. That affects the screenshot too.

**Review by sussexrick — CHANGES_REQUESTED** [review]

File: docs/components/text-input.md (hunk 1 above)
Comment (samuelhaa) [reply docs/components/text-input.md]: Changed the heading to h1 and renamed it Text input

File: GovUk.Frontend.AspNetCore.Extensions/Validation/RegisteredCharityNumberAttribute.cs (hunk 3 above)
Comment (samuelhaa) [reply GovUk.Frontend.AspNetCore.Extensions/Validation/RegisteredCharityNumberAttribute.cs]: Updated the default error message and screenshot

**Review by sussexrick — APPROVED** [review]

**Pull request merged**